From: John Crispin <john@phrozen.org>
Subject: [PATCH] Makefile: honour SOURCE_DATE_EPOCH for the build timestamp

The rule for timestamp_autogenerated.h calls date with no argument, so
U_BOOT_DATE and U_BOOT_TIME hold the wall clock. The banner of every
u-boot image therefore changes on each build:

  U-Boot 2013.10-openwrt4 (Sep 21 2026 - 20:47:50) ARV7506PW11

Use SOURCE_DATE_EPOCH when the environment provides it, as described in
https://reproducible-builds.org/specs/source-date-epoch/, and keep the
wall clock otherwise. Later u-boot versions do the same in their own
timestamp rule.

Signed-off-by: John Crispin <john@phrozen.org>
---
 Makefile | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

--- a/Makefile
+++ b/Makefile
@@ -15,6 +15,9 @@
 U_BOOT_VERSION = $(VERSION).$(PATCHLEVEL)$(EXTRAVERSION)
 endif
 TIMESTAMP_FILE = $(obj)include/generated/timestamp_autogenerated.h
+ifneq ($(SOURCE_DATE_EPOCH),)
+SOURCE_DATE_ARGS = -u -d @$(SOURCE_DATE_EPOCH)
+endif
 VERSION_FILE = $(obj)include/generated/version_autogenerated.h

 HOSTARCH := $(shell uname -m | \
@@ -804,8 +807,8 @@

 $(TIMESTAMP_FILE):
 		@mkdir -p $(dir $(TIMESTAMP_FILE))
-		@LC_ALL=C date +'#define U_BOOT_DATE "%b %d %C%y"' > $@.tmp
-		@LC_ALL=C date +'#define U_BOOT_TIME "%T"' >> $@.tmp
+		@LC_ALL=C date $(SOURCE_DATE_ARGS) +'#define U_BOOT_DATE "%b %d %C%y"' > $@.tmp
+		@LC_ALL=C date $(SOURCE_DATE_ARGS) +'#define U_BOOT_TIME "%T"' >> $@.tmp
 		@cmp -s $@ $@.tmp && rm -f $@.tmp || mv -f $@.tmp $@

 easylogo env gdb:
